apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ir/IRBuilder.kt

```diff
@@ -0,0 +1,258 @@
+
+import com.apollographql.apollo.compiler.ApolloMetadata
+import com.apollographql.apollo.compiler.applyIf
+import com.apollographql.apollo.compiler.ir.TypeDeclaration.Companion.KIND_ENUM
+import com.apollographql.apollo.compiler.ir.TypeDeclaration.Companion.KIND_INPUT_OBJECT_TYPE
+import com.apollographql.apollo.compiler.ir.TypeDeclaration.Companion.KIND_SCALAR_TYPE
+import com.apollographql.apollo.compiler.parser.error.DocumentParseException
+import com.apollographql.apollo.compiler.parser.error.ParseException
+import com.apollographql.apollo.compiler.parser.graphql.DocumentParseResult
+import com.apollographql.apollo.compiler.parser.graphql.checkMultipleFragmentDefinitions
+import com.apollographql.apollo.compiler.parser.graphql.checkMultipleOperationDefinitions
+import com.apollographql.apollo.compiler.parser.graphql.validateArguments
+import com.apollographql.apollo.compiler.parser.introspection.IntrospectionSchema
+import com.apollographql.apollo.compiler.parser.introspection.asGraphQLType
+import com.apollographql.apollo.compiler.parser.introspection.possibleTypes
+
+class IRBuilder(private val schema: IntrospectionSchema,
+                private val schemaPackageName: String,
+                private val incomingMetadata: ApolloMetadata?,
+                private val alwaysGenerateTypesMatching: Set<String>?,
+                generateMetadata: Boolean
+) {
+  private val isRootCompilationUnit = incomingMetadata == null && generateMetadata
+
+  private fun extraTypes(): Set<String> {
+    val regexes = (alwaysGenerateTypesMatching ?: (listOf(".*").takeIf { isRootCompilationUnit } ?: emptyList()))
+        .map { Regex(it) }
+
+    return schema.types.values.filter { type ->
+      (type.kind == IntrospectionSchema.Kind.ENUM
+          || type.kind == IntrospectionSchema.Kind.INPUT_OBJECT)
+          && regexes.indexOfFirst { it.matches(type.name) } >= 0
+    }.map { it.name }
+        .toSet()
+  }
+
+  fun build(documentParseResult: DocumentParseResult): CodeGenerationIR {
+    val allFragments = (incomingMetadata?.fragments ?: emptyList()) + documentParseResult.fragments
+
+    documentParseResult.operations.checkMultipleOperationDefinitions()
+    allFragments.checkMultipleFragmentDefinitions()
+
+    val fragmentsToGenerate = documentParseResult.fragments.map { it.fragmentName }
+
+    val incomingTypes = incomingMetadata?.types ?: emptySet()
+    val extraTypes = extraTypes()
+
+    val typeDeclarations = (documentParseResult.usedTypes + extraTypes) .usedTypeDeclarations()
+
+    val enumsToGenerate = typeDeclarations.filter { it.kind == KIND_ENUM }
+        .map { it.name }
+        .filter { !incomingTypes.contains(it) }
+
+    val inputObjectsToGenerate = typeDeclarations.filter { it.kind == KIND_INPUT_OBJECT_TYPE }
+        .map { it.name }
+        .filter { !incomingTypes.contains(it) }
+
+    // Always generate the extra "ID" scalar type
+    // I'm not 100% sure why this is required but keep this for backward compatibility
```

Comment:
Do we need to generate "ID" in the `CustomType` enum? Is there a use case where users can register a custom adapter for "ID" ?